### PR TITLE
Add CxPlatAsync to simplify creating threads with cxplat

### PR DESCRIPTION
**Pull request by csujedihy**
One can easily spin off a thread similar to std::async with this class.

```
CxPlatAsync Async([](void*) -> void* { /* do something */ return nullptr; });
Async.Wait();

Async.Get(); // get the return value from the callback
```

Lambda capture could be supported but my C++ skill is not sufficient.

### CODE REVIEW

File: src/test/CxPlatTests.h [hunk 1]
```diff
@@ -135,10 +136,13 @@ static const GUID CXPLAT_TEST_DEVICE_INSTANCE =
 #define IOCTL_CXPLAT_RUN_THREAD_BASIC \
     CXPLAT_CTL_CODE(6, METHOD_BUFFERED, FILE_WRITE_DATA)
 
-#define IOCTL_CXPLAT_RUN_THREAD_WAIT_TIMEOUT \
+#define IOCTL_CXPLAT_RUN_THREAD_ASYNC \
```
Comment (nibanks) [inline src/test/CxPlatTests.h]: It'd make the diff nicer to make `IOCTL_CXPLAT_RUN_THREAD_ASYNC ` number 9 (last) instead.

File: inc/cxplat.hpp [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    struct CxPlatAsyncContext {
+        void* UserContext;
+        CxPlatCallback *UserCallback;
+        void* ReturnValue;
+    };
+
+    static
+    CXPLAT_THREAD_CALLBACK(CxPlatAsyncWrapperCallback, Context)
+    {
+        struct CxPlatAsyncContext *AsyncContext = (struct CxPlatAsyncContext*)Context;
```
Comment (nibanks) [inline inc/cxplat.hpp]:
```suggestion
        struct CxPlatAsyncContext* AsyncContext = (struct CxPlatAsyncContext*)Context;
```

File: inc/cxplat.hpp [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+class CxPlatAsync {
+private:
+    struct CxPlatAsyncContext {
+        void* UserContext;
+        CxPlatCallback *UserCallback;
+        void* ReturnValue;
+    };
+
+    static
+    CXPLAT_THREAD_CALLBACK(CxPlatAsyncWrapperCallback, Context)
```
Comment (nibanks) [inline inc/cxplat.hpp]:
```suggestion
    static CXPLAT_THREAD_CALLBACK(CxPlatAsyncWrapperCallback, Context)
```

File: inc/cxplat.hpp [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    CXPLAT_THREAD Thread {0};
+    CXPLAT_THREAD_CONFIG ThreadConfig {0};
+    struct CxPlatAsyncContext AsyncContext {0};
+    bool Initialized = false;
+    bool ThreadCompleted = false;
+public:
+    CxPlatAsync(CxPlatCallback Callback, void* UserContext = nullptr) noexcept {
+        AsyncContext.UserContext = UserContext;
+        AsyncContext.UserCallback = Callback;
+        AsyncContext.ReturnValue = nullptr;
```
Comment (nibanks) [inline inc/cxplat.hpp:50]: Should be unnecessary, right?

File: inc/cxplat.hpp [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    }
+
+    void Wait() noexcept {
+        if (Initialized) {
+            CxPlatThreadWaitForever(&Thread);
+            ThreadCompleted = true;
+        }
+    }
+
+#if defined(CX_PLATFORM_WINUSER) || defined(CX_PLATFORM_WINKERNEL)
```
Comment (nibanks) [inline inc/cxplat.hpp:77]: Why not Posix?

File: inc/cxplat.hpp [hunk 6]
```diff
@@ -0,0 +1,87 @@
+/*++
+
+    Copyright (c) Microsoft Corporation.
+    Licensed under the MIT License.
+
+Abstract:
+
+    C++ header only cxplat wrappers
+
+--*/
```
Comment (nibanks) [inline inc/cxplat.hpp:10]: Add

```c
#ifdef _WIN32
#pragma once
#endif
```

before `#ifndef CXPLATCPP_H`

File: inc/cxplat.hpp (hunk 4 above)
Comment (csujedihy) [reply inc/cxplat.hpp:50]: They are necessary to hide the thread callback signature difference between all platforms.

File: inc/cxplat.hpp (hunk 5 above)
Comment (csujedihy) [reply inc/cxplat.hpp:77]: We have not implemented this functionality in both msquic and cxplat.

File: inc/cxplat.hpp (hunk 6 above)
Comment (csujedihy) [reply inc/cxplat.hpp:10]: why is this necessary though? `#ifndef CXPLATCPP_H` already guarantees that the header is used once. Does `pragma` make compile faster or something?

Comment (nibanks) [reply inc/cxplat.hpp:10]: pragmas are faster AFAIU

**Review by nibanks — APPROVED** [review]

**Review by nibanks — APPROVED** [review]

**Review by nibanks — APPROVED** [review]

**Comment by csujedihy** [comment]
I gave up on implementing thread wait with a timeout. pthread_timedjoin_np doesn't exist in android and mac osx.

**Pull request merged**